In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [13]:
data = pd.read_csv("./data/heart.csv")

In [14]:
data.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1


In [4]:
data.dtypes

age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca            int64
thal          int64
target        int64
dtype: object

# Giải thích các đặc trưng có trong bảng
### 1. Age: Tuổi của bệnh nhân
### 2. Sex: Giới tính
    Value 0: female
    Value 1: male
### 3. cp: Loại đau ngực
    Value 0: Đau thắt ngực
    Value 1: Đau thắt ngực không điển hình
    Value 2: Đau không phải do đau thắt ngực
    Value 3: Không có triệu chứng
### 4. trestbps: Huyết áp khi nghỉ ngơi 
### 5. chol: Nồng độ cholestoral trong huyết thanh (mg/dl)
### 6. fbs: Đường huyết lúc đói > 120 mg/dl
    Value 0: false
    Value 1: true
### 7. restecg: Kết quả điện tâm đồ lúc nghỉ ngơi
    Value 0: bình thườngS
    Value 1: có bất thường sóng ST-T (sóng T đảo ngược và/hoặc ST chênh lên hoặc chênh xuống > 0,05 mV)
    Value 2: cho thấy có khả năng hoặc chắc chắn bị phì đại thất trái theo tiêu chí của Estes.
### 8. thalach: Nhịp tim tối đa đạt đượcK
### 9. exang: Đau thắt ngực khi tập thể dục
    Value 0: No
    Value 1: Yes
### 10. oldpeak: Độ chênh xuống đoạn ST khi gắng sứcB
### 11. slope: Độ dốc của đoạn ST khi tập luyện ở mức cao nhất
### 12. ca: Số lượng mạch máu chính (0-3) được nhuộm màu bằng huỳnh quang
### 13. thal:
    Value 0: Lỗi giá trị (Giá trị naỳ không có ý nghĩa)
    Value 1: Khiếm khuyết cố định
    Value 2: Bình thường
    Value 3: Khiếm khuyết có thể hồi phụ
### 14. target: Mục tiêu 
    Value 0: Không mắc bệnh tim
    Value 1: Mắc bệnh tim

In [15]:
#Check error thal data
check1 = data[data["thal"] == 0]
check1.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
48,53,0,2,128,216,0,0,115,0,0.0,2,0,0,1
281,52,1,0,128,204,1,1,156,1,1.0,1,0,0,0


In [22]:
#Check error ca data 
heck2 = data[data["ca"] > 3]
check2.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
92,52,1,2,138,223,0,1,169,0,0.0,2,4,2,1
158,58,1,1,125,220,0,1,144,0,0.4,1,4,3,1
163,38,1,2,138,175,0,1,173,0,0.0,2,4,2,1
164,38,1,2,138,175,0,1,173,0,0.0,2,4,2,1
251,43,1,0,132,247,1,0,143,1,0.1,1,4,3,0


In [18]:
data.shape

(303, 14)

In [23]:
#Delete wrong data
data = data[data["thal"] > 0]
data = data [data["ca"] < 4]
data.shape

(296, 14)

In [24]:
#Check missing columns data
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 296 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       296 non-null    int64  
 1   sex       296 non-null    int64  
 2   cp        296 non-null    int64  
 3   trestbps  296 non-null    int64  
 4   chol      296 non-null    int64  
 5   fbs       296 non-null    int64  
 6   restecg   296 non-null    int64  
 7   thalach   296 non-null    int64  
 8   exang     296 non-null    int64  
 9   oldpeak   296 non-null    float64
 10  slope     296 non-null    int64  
 11  ca        296 non-null    int64  
 12  thal      296 non-null    int64  
 13  target    296 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 34.7 KB


In [26]:
data.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

In [28]:
data.describe()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
count,296.000000,296.000000,296.000000,296.00000,296.000000,296.00000,296.000000,296.000000,296.000000,296.000000,296.000000,296.000000,296.000000,296.000000
mean,54.523649,0.679054,0.959459,131.60473,247.155405,0.14527,0.523649,149.560811,0.327703,1.059122,1.395270,0.679054,2.327703,0.540541
std,9.059471,0.467631,1.034184,17.72662,51.977011,0.35297,0.526692,22.970792,0.470171,1.166474,0.618235,0.939726,0.585743,0.499198
min,29.000000,0.000000,0.000000,94.00000,126.000000,0.00000,0.000000,71.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000
25%,48.000000,0.000000,0.000000,120.00000,211.000000,0.00000,0.000000,133.000000,0.000000,0.000000,1.000000,0.000000,2.000000,0.000000
50%,56.000000,1.000000,1.000000,130.00000,242.500000,0.00000,1.000000,152.500000,0.000000,0.800000,1.000000,0.000000,2.000000,1.000000
75%,61.000000,1.000000,2.000000,140.00000,275.250000,0.00000,1.000000,166.000000,1.000000,1.650000,2.000000,1.000000,3.000000,1.000000
max,77.000000,1.000000,3.000000,200.00000,564.000000,1.00000,2.000000,202.000000,1.000000,6.200000,2.000000,3.000000,3.000000,1.000000


In [33]:
#Convert data types 
features = ["sex", "cp", "fbs", "restecg", "exang", "slope", "thal"]
for i in features:
    data[i] = data[i].astype("category")

In [35]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 296 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   age       296 non-null    int64   
 1   sex       296 non-null    category
 2   cp        296 non-null    category
 3   trestbps  296 non-null    int64   
 4   chol      296 non-null    int64   
 5   fbs       296 non-null    category
 6   restecg   296 non-null    category
 7   thalach   296 non-null    int64   
 8   exang     296 non-null    category
 9   oldpeak   296 non-null    float64 
 10  slope     296 non-null    category
 11  ca        296 non-null    int64   
 12  thal      296 non-null    category
 13  target    296 non-null    int64   
dtypes: category(7), float64(1), int64(6)
memory usage: 21.5 KB


In [37]:
data.describe()

,age,trestbps,chol,thalach,oldpeak,ca,target
count,296.000000,296.00000,296.000000,296.000000,296.000000,296.000000,296.000000
mean,54.523649,131.60473,247.155405,149.560811,1.059122,0.679054,0.540541
std,9.059471,17.72662,51.977011,22.970792,1.166474,0.939726,0.499198
min,29.000000,94.00000,126.000000,71.000000,0.000000,0.000000,0.000000
25%,48.000000,120.00000,211.000000,133.000000,0.000000,0.000000,0.000000
50%,56.000000,130.00000,242.500000,152.500000,0.800000,0.000000,1.000000
75%,61.000000,140.00000,275.250000,166.000000,1.650000,1.000000,1.000000
max,77.000000,200.00000,564.000000,202.000000,6.200000,3.000000,1.000000


In [39]:
#Split 80% train and 20% test
from sklearn.model_selection import train_test_split
x = data.drop("target", axis=1)
y = data["target"]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)